### PR TITLE
feature/db/sub_repo_perms: add IPs column to sub repo perms

### PR DESCRIPTION
**Pull request by ggilmore**
This PR modifies the sub_repo_permissions database to store the ip addresses associated with each Perforce path rule, as part of the IP based sub repo permissions work. 

The new IP column is implemetned as a []text similar to the path column. The IP address associated with `paths[0]` is stored in the ips column in `ips[0]`. 

For example, the follownig proections table 

```
Protections:
    read     user     emily   *                    //depot/elm_proj/...
    write    group    devgrp  *                    //...
    write    user     *       192.168.41.0/24     -//...
    write    user     *       [2001:db8:1:2::]/64 -//...
    write    user     joe     *                   -//...
    write    user     lisag   *                   -//depot/...
    write    user     lisag   *                    //depot/doc/...
    super    user     edk     *                    //...
```

turns into the following rows in the sub_repo_permissions table

| repo_id | user_id | version | updated_at | paths | ips |
|---------|---------|---------|------------|-------|-----|
| 1 | 1 | 1 | 2023-07-01 10:00:00 | {`"//depot/elm_proj/..."`} | {`"*"`} |
| 1 | 2 | 1 | 2023-07-01 10:00:00 | {`"//..."`} | {`"*"`} |
| 1 | 3 | 1 | 2023-07-01 10:00:00 | {`"-//..."`} | {`"192.168.41.0/24"`} |
| 1 | 4 | 1 | 2023-07-01 10:00:00 | {`"-//..."`} | {`"2001:db8:1:2::]/64"`} |
| 1 | 5 | 1 | 2023-07-01 10:00:00 | {`"-//..."`} | {`"*"`} |
| 1 | 6 | 1 | 2023-07-01 10:00:00 | {`"-//depot/..."`, `"//depot/doc/..."`} | {`"*"`, `"*"`} |
| 1 | 7 | 1 | 2023-07-01 10:00:00 | {`"//..."`} | {`"*"`} |

## Test plan

The unit test for the sub_repository_permissions store PR that is built on this PR: https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63811/internal-database-sub_repo_permissions-modify-store-to-be-able-to-insert-ip-based-permissions

## Changelog

- The sub_repo_permissions table now has an ips column to store the associated IP address associated with each path rule.

### CODE REVIEW

**Comment by ggilmore** [comment]
* **#63771** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63771?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ggilmore and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/database/schema.md [hunk 1]
```diff
@@ -3989,9 +3989,12 @@ Foreign-key constraints:
  version    | integer                  |           | not null | 1
  updated_at | timestamp with time zone |           | not null | now()
  paths      | text[]                   |           |          | 
+ ips        | text[]                   |           |          | 
 Indexes:
     "sub_repo_permissions_repo_id_user_id_version_uindex" UNIQUE, btree (repo_id, user_id, version)
     "sub_repo_perms_user_id" btree (user_id)
+Check constraints:
+    "ips_paths_length_check" CHECK (ips IS NULL OR array_length(ips, 1) = array_length(paths, 1) AND NOT (''::text = ANY (ips)))
```
Comment (ggilmore) [inline internal/database/schema.md:3997]: This ensures

- that we have an IP address associated with each path entry, or no ips at all (if the perforce permissions syncer hasn't been updated yet to save those)
- that we have no blank IP address entries (this doesn't seem to be possible in perforce, so there is no reason to allow it here either)

File: migrations/frontend/1720651147_add_ips_to_sub_repo_perms_table/up.sql [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+-- See /migrations/README.md. Highlights:
+--  * Make migrations idempotent (use IF EXISTS)
+--  * Make migrations backwards-compatible (old readers/writers must continue to work)
+--  * If you are using CREATE INDEX CONCURRENTLY, then make sure that only one statement
+--    is defined per file, and that each such statement is NOT wrapped in a transaction.
+--    Each such migration must also declare "createIndexConcurrently: true" in their
+--    associated metadata.yaml file.
+--  * If you are modifying Postgres extensions, you must also declare "privileged: true"
+--    in the associated metadata.yaml file.
+
```
Comment (eseliger) [inline migrations/frontend/1720651147_add_ips_to_sub_repo_perms_table/up.sql:12]: I don't think we need to keep the instructions 

```suggestion
```

File: migrations/frontend/1720651147_add_ips_to_sub_repo_perms_table/down.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+-- Undo the changes made in the up migration
+
```
Comment (eseliger) [inline migrations/frontend/1720651147_add_ips_to_sub_repo_perms_table/down.sql:2]:
```suggestion
```

**Review by eseliger — COMMENTED** [review]
Needs a test plan, the auditor wrongly marked this as ok.

**Review by eseliger — APPROVED** [review]

**Comment by ggilmore** [comment]
### Merge activity

* **Jul 18, 3:54 PM EDT**: @ggilmore merged this pull request with [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63771).

**Pull request merged**